### PR TITLE
Opencl beta proportion lpdf

### PR DESCRIPTION
**Pull request by t4c1**
## Summary

Adds an OpenCL implementation of `beta_proportion_lpdf` function.

## Tests

Added tests.

## Side Effects

None.

## Release notes

Added an OpenCL implementation of `beta_proportion_lpdf` function.

## Checklist

- [ ] Math issue #2152 

- [ ] Copyright holder: Tadej Ciglarič

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [ ] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [ ] the code is written in idiomatic C++ and changes are documented in the doxygen

- [ ] the new changes are tested

### CODE REVIEW

**Comment by bbbales2** [comment]
@t4c1 yo I've done the probability reviews up to here but I don't know how the OpenCL stuff works so you'll need to grab someone else from here on (which is why I haven't reviewed this).

**Comment by t4c1** [comment]
Yeah, that is fine. @rok-cesnovar will review these. There is no rush here.

**Review by rok-cesnovar — CHANGES_REQUESTED** [review]
A couple of questions, but otherwise looks good.

File: stan/math/opencl/prim/beta_proportion_lpdf.hpp [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  auto check_mu_bounded = check_cl(function, "Location parameter", mu_val,
+                                   "in the interval (0, 1)");
+  auto mu_bounded_expr = 0 < mu_val && mu_val < 1;
+  auto check_kappa_positive_finite = check_cl(
+      function, "Precision parameter", kappa_val, "in the interval [0, 1]");
+  auto kappa_positive_finite = 0 < kappa_val && isfinite(kappa_val);
+
+  auto log_y_expr = log(y_val);
+  auto log1m_y_expr = log1p(-y_val);
+  auto mukappa_expr = elt_multiply(mu_val, kappa_val);
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/opencl/prim/beta_proportion_lpdf.hpp:74]: In Eigen version matrices are converted to arrays, so * is the same as elt_multiply

File: stan/math/opencl/prim/beta_proportion_lpdf.hpp [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+  const auto& y_val = value_of(y);
+  const auto& mu_val = value_of(mu);
+  const auto& kappa_val = value_of(kappa);
+
+  auto check_y_bounded
+      = check_cl(function, "Random variable", y_val, "in the interval [0, 1]");
+  auto y_bounded_expr = 0 <= y_val && y_val <= 1;
+  auto check_mu_bounded = check_cl(function, "Location parameter", mu_val,
+                                   "in the interval (0, 1)");
+  auto mu_bounded_expr = 0 < mu_val && mu_val < 1;
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/opencl/prim/beta_proportion_lpdf.hpp:67]: In prim/prob this one actually uses `check_less` and `check_positive`, which do not allow boundary values.

File: stan/math/opencl/prim/beta_proportion_lpdf.hpp [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+  auto check_kappa_positive_finite = check_cl(
+      function, "Precision parameter", kappa_val, "in the interval [0, 1]");
+  auto kappa_positive_finite = 0 < kappa_val && isfinite(kappa_val);
+
+  auto log_y_expr = log(y_val);
+  auto log1m_y_expr = log1p(-y_val);
```
[the remark this answers is not in the document]
Comment (t4c1) [reply stan/math/opencl/prim/beta_proportion_lpdf.hpp:73]: We don't OpenCL log1m, although I could add it.

[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/opencl/prim/beta_proportion_lpdf.hpp:73]: Oh, not necessary just for this case.

File: stan/math/opencl/prim/beta_proportion_lpdf.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (rok-cesnovar) [reply stan/math/opencl/prim/beta_proportion_lpdf.hpp:67]: Ah yes, I mixed it up with y_val. We are cool.

**Review by rok-cesnovar — APPROVED** [review]

**Pull request merged**